### PR TITLE
GT-1561, expose created_at, tweak paths for user and counters to support "me"

### PR DESCRIPTION
**Pull request by andrewroth**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -30,6 +30,12 @@
 
   resources :custom_manifests, only: [:create, :update, :destroy, :show]
 
+  patch "user/counters/:id", to: "user_counters#update"    # Legacy route for GodTools Android v5.7.0-v6.0.0
+  patch "user/me/counters/:id", to: "user_counters#update" # Legacy route for GodTools Android v6.0.1+
```
Comment (frett) [inline config/routes.rb:34]: how do these interact with the `scope "user"` block right below it? show the block below this be removed?

File: app/controllers/with_user_controller.rb [hunk 2]
```diff
@@ -7,6 +7,13 @@ def authorize!
...
     return if authorization && current_user
 
+    # currently, if trying to operate on a specific @user, you can only operate on your own user data, but this may change later
+    return if @user && @user == current_user
+
     render_unauthorized
   end
+
+  def set_user
+    @user = params[:user_id].blank? || params[:user_id] == "me" ? current_user : User.find(params[:user_id])
```
Comment (knutsenm) [inline app/controllers/with_user_controller.rb:17]: So long as rubocop doesn't complain, I would appreciate parens here:
```suggestion
    @user = (params[:user_id].blank? || params[:user_id] == "me") ? current_user : User.find(params[:user_id])
```

File: app/controllers/with_user_controller.rb [hunk 3]
```diff
@@ -7,6 +7,13 @@ def authorize!
     # requested is authorized if using okta to provide a valid user id
     return if authorization && current_user
 
+    # currently, if trying to operate on a specific @user, you can only operate on your own user data, but this may change later
+    return if @user && @user == current_user
```
Comment (knutsenm) [inline app/controllers/with_user_controller.rb:11]: Your new UsersController does not `set_user`; will this behave properly there?

File: app/serializers/user_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class UserSerializer < ActiveModel::Serializer
+  type "user"
+  attributes :sso_guid, :created_at
+
+  def created_at
+    object.created_at.iso8601 # without this, the default serializer datetime will add 3 ms digits which Daniel prefers not to have
```
Comment (knutsenm) [inline app/serializers/user_serializer.rb:8]: The royal "we" will suffice instead of "Daniel" ;-)

**Review by knutsenm — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 1 above)
Comment (andrewroth) [reply config/routes.rb:34]: Oh, yeah, the scope stuff can be removed

File: app/controllers/with_user_controller.rb (hunk 3 above)
Comment (andrewroth) [reply app/controllers/with_user_controller.rb:11]: Right, I realized that and did some more work on this, I'll re-request the review

File: app/controllers/with_user_controller.rb [hunk 5]
```diff
@@ -4,9 +4,16 @@ class WithUserController < ApplicationController
...
+    return if authorization && current_user && !@user
+
+    # currently, if trying to operate on a specific @user, you can only operate on your own user data, but this may change later
+    return if @user && @user == current_user
 
     render_unauthorized
   end
+
+  def set_user(user_id_attribute = :user_id)
+    @user = params[user_id_attribute].blank? || params[user_id_attribute] == "me" ? current_user : User.find(params[user_id_attribute])
```
Comment (knutsenm) [inline app/controllers/with_user_controller.rb:17]: So long as rubocop doesn't complain, I would appreciate parens here:
```suggestion
    @user = (params[user_id_attribute].blank? || params[user_id_attribute] == "me") ? current_user : User.find(params[user_id_attribute])
```

File: config/routes.rb [hunk 6]
```diff
@@ -30,23 +30,21 @@
...
   resources :custom_manifests, only: [:create, :update, :destroy, :show]
 
-  scope "user" do
-    resources :counters, controller: "user_counters", only: [:update] # Legacy route for GodTools Android v5.7.0-v6.0.0
-
-    scope "me" do
-      resources :counters, controller: "user_counters", only: [:index, :update]
-    end
-  end
+  patch "user/counters/:id", to: "user_counters#update"    # Legacy route for GodTools Android v5.7.0-v6.0.0
```
Comment (knutsenm) [inline config/routes.rb:33]: Rubocop is being even more pedantic than me :-/ - doesn't like the comment spacing

**Review by knutsenm — CHANGES_REQUESTED** [review]

**Review by knutsenm — APPROVED** [review]

**Review by frett — APPROVED** [review]
everything appears to be functioning correctly on stage

File: app/controllers/with_user_controller.rb (hunk 5 above)
Comment (andrewroth) [reply app/controllers/with_user_controller.rb:17]: Sorry, no such luck.

standard: Use Ruby Standard Style (https://github.com/testdouble/standard)
standard: Run `standardrb --fix` to automatically fix some problems.
  app/controllers/with_user_controller.rb:17:13: Style/TernaryParentheses: Omit parentheses for ternary conditions.

I like the look of parens here also.. :shrug:

**Pull request merged**